internal/uvm/vpmem.go

```diff
@@ -2,60 +2,301 @@ package uvm
+	}).Debug("mapped new device")
+
+	vpd.mappings[device.HostPath()] = device
+	return nil
+}
+
+// unmapDevice reduces reference of a specified device and takes care of releasing the allocated
+// memory when appropriate
+func (vpd *vpmemDevice) unmapDevice(ctx context.Context, hostPath string) error {
+	dev, ok := vpd.mappings[hostPath]
+	if !ok {
+		return ErrNotAttached
+	}
+
+	if dev.RefCount() == 1 {
+		log.G(ctx).WithFields(logrus.Fields{
+			"hostPath": hostPath,
+		}).Debug("releasing resources")
+		memCls := getMemoryClassType(dev.DeviceSize())
+		err := vpd.release(memCls, dev.DeviceOffset())
+		if err != nil {
+			return err
+		}
+		delete(vpd.mappings, hostPath)
+		return nil
+	}
+	dev.RemoveRef()
+	return nil
+}
+
+// addMappedDevice creates a new vpmemDevice and adds a mapping to it
+func (uvm *UtilityVM) addMappedDevice(ctx context.Context, device mappedDevice) error {
+	pmem := uvm.vpmemDevices[device.DeviceNumber()]
+	if pmem == nil {
+		pmem = newVPMemDevice()
+		uvm.vpmemDevices[device.DeviceNumber()] = pmem
+	}
+
+	return pmem.mapDevice(ctx, device)
+}
+
+// NewHostComputeRequest generates host and guest modification request which is passed
+// to the host compute service
+func (uvm *UtilityVM) NewHostComputeRequest(ctx context.Context, rType string, device refDevice) (request *hcsschema.ModifySettingRequest, err error) {
+	log.G(ctx).WithFields(logrus.Fields{
+		"requestType": rType,
+		"device":      device.DeviceNumber(),
+		"hostPath":    device.HostPath(),
+		"mountPath":   device.MountPath(),
+	}).Debug("generating hcs request")
+
+	request = &hcsschema.ModifySettingRequest{
+		RequestType: rType,
+	}
+
+	dev := uvm.vpmemDevices[device.DeviceNumber()]
+	if rType == requesttype.Add {
+		if dev == nil {
+			request.Settings = hcsschema.VirtualPMemDevice{
+				ReadOnly:    true,
+				HostPath:    device.HostPath(),
+				ImageFormat: "Vhd1",
+			}
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, device.DeviceNumber())
+		} else {
+			request.Settings = device.HostSettings()
+			request.ResourcePath = device.ResourcePath()
+		}
+	} else if rType == requesttype.Remove {
+		if len(dev.mappings) == 1 {
+			request.ResourcePath = fmt.Sprintf(vPMemControllerResourceFormat, device.DeviceNumber())
+		} else {
+			request.ResourcePath = device.ResourcePath()
+		}
+	} else {
+		return nil, errors.New("unsupported request type")
+	}
+
+	request.GuestRequest = guestrequest.GuestRequest{
+		ResourceType: guestrequest.ResourceTypeVPMemDevice,
+		RequestType:  rType,
+		Settings:     device.GuestSettings(),
+	}
+
+	return request, nil
+}
+
 // findNextVPMEM finds the next available VPMem slot.
 //
 // The lock MUST be held when calling this function.
-func (uvm *UtilityVM) findNextVPMEM(ctx context.Context, hostPath string) (uint32, error) {
+func (uvm *UtilityVM) findNextVPMEM(hostPath string) (dev mappedDevice, err error) {
+	var fi os.FileInfo
+	fi, err = os.Stat(hostPath)
+	if err != nil {
+		return nil, err
+	}
```

Comment:
We should normalize the `hostPaths` given so we don't try to map multiple things that are actually the same.